src/containers/qset.c

```diff
@@ -0,0 +1,530 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES {} LOSS OF USE, DATA, OR PROFITS {} OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *****************************************************************************/
+
+/**
+ * @file qset.c Mathematical Set Implementation
+ * qset container is a mathemetical set implementation.
+ *
+ */
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <stdbool.h>
+#include "qinternal.h"
+#include "containers/qset.h"
+
+#define MAX_FULLNESS_PERCENT 0.25
+
+#ifndef _DOXYGEN_SKIP
+
+static uint64_t __default_hash(const char *key);
+static bool __get_index(qset_t *set, const char *key, uint64_t hash, uint64_t *index);
+static bool __assign_node(qset_t *set, const char *key, uint64_t hash, uint64_t index);
+static void __free_index(qset_t *set, uint64_t index);
+static bool __set_contains(qset_t *set, const char *key, uint64_t hash);
+static bool __set_add(qset_t *set, const char *key, uint64_t hash);
+static void __relayout_nodes(qset_t *set, uint64_t start, short end_on_null);
+
+#endif
+
+/**
+ * Create new qset_t container
+ * 
+ * @param num_els num of elements
+ * @param hash hash function callback
+ * @param options combination of initialization options.
+ * 
+ * @return a pointer of malloced qset_t container, otherwise returns NULL
+ * @retval errno will be set in error condition.
+ *   - ENOMEM and QSET_MALLERR : Malloc failure.
+ *   - EINVAL : Invalid argument.
+ * 
+ * @code
+ *   qset_t *set = qset(10, hash, 0);
+ * @endcode
+ * 
+ * @note
+ *   Available options:
+ *   - QSET_THREADSAFE - make it thread-safe.
+ * 
+ */
+qset_t *qset(uint64_t num_els, qset_hashfunction_t hash, int options) {
+    if(num_els == 0) {
+        ret_status = QSET_MEMERR;
+        return NULL;
+    }
+    qset_t *set = (qset_t *) calloc(1, sizeof(qset_t));
```

Comment:
Would it be possible for qset to use one of existing containers as a base container such as qlist?

Reply:
nope. Cuz qset_object_t has two things